src/main/java/com/mongodb/jdbc/oidc/RFC8252HttpServer.java

```diff
@@ -0,0 +1,256 @@
+                Context context = new Context();
+                context.setVariable(ERROR_URI_KEY, LOGIN_ERROR_URI);
+                context.setVariable(PRODUCT_DOCS_LINK_KEY, PRODUCT_DOCS_LINK);
+                context.setVariable(PRODUCT_DOCS_NAME_KEY, PRODUCT_DOCS_NAME);
+                context.setVariable(ERROR_KEY, queryParams.get(ERROR_KEY));
+                context.setVariable(
+                        ERROR_DESCRIPTION_KEY,
+                        queryParams.getOrDefault(ERROR_DESCRIPTION_KEY, "Unknown error"));
+                String errorHtml = templateEngine.process("OIDCErrorTemplate", context);
+                sendResponse(exchange, errorHtml, HttpURLConnection.HTTP_BAD_REQUEST);
+
+            } else {
+                oidcResponse.setError("Not found");
+                String allParams =
+                        queryParams
+                                .entrySet()
+                                .stream()
+                                .map(entry -> entry.getKey() + "=" + entry.getValue())
+                                .reduce((param1, param2) -> param1 + ", " + param2)
+                                .orElse("No parameters");
+                oidcResponse.setErrorDescription("Not found. Parameters: " + allParams);
+                if (!putOidcResponse(exchange, oidcResponse)) {
+                    return;
+                }
+                Context context = new Context();
+                context.setVariable(PRODUCT_DOCS_LINK_KEY, PRODUCT_DOCS_LINK);
+                context.setVariable(PRODUCT_DOCS_NAME_KEY, PRODUCT_DOCS_NAME);
+                String notFoundHtml = templateEngine.process("OIDCNotFoundTemplate", context);
+                sendResponse(exchange, notFoundHtml, HttpURLConnection.HTTP_NOT_FOUND);
+            }
+            exchange.close();
+        }
+    }
+
+    /** HTTP handler for handling the accepted endpoint. */
+    private class AcceptedHandler implements HttpHandler {
+        @Override
+        public void handle(HttpExchange exchange) throws IOException {
+            Context context = new Context();
+            context.setVariable(PRODUCT_DOCS_LINK_KEY, PRODUCT_DOCS_LINK);
+            context.setVariable(PRODUCT_DOCS_NAME_KEY, PRODUCT_DOCS_NAME);
+            String acceptedHtml = templateEngine.process("OIDCAcceptedTemplate", context);
+            sendResponse(exchange, acceptedHtml, HttpURLConnection.HTTP_OK);
+            exchange.close();
+        }
+    }
+
+    /**
+     * Parses the query parameters from the HTTP exchange.
+     *
+     * @param exchange the HTTP exchange
+     * @return a map containing the parsed query parameters
+     * @throws UnsupportedEncodingException if the encoding is not supported
+     */
+    private Map<String, String> parseQueryParams(HttpExchange exchange)
+            throws UnsupportedEncodingException {
+        Map<String, String> queryParams = new HashMap<>();
+        String query = exchange.getRequestURI().getQuery();
```

Comment:
If there are any `&` or `=` as part of the values passed in (we know it shouldn't be part of the keys at least) then the logic you have using split below won't work.
You need to use getRawQuery, especially since you are decoding the values already when processing the params, you are already all set.

Reply:
I made this change and added some test cases for having a `&` or `=` in the value.  Good catch.